### ISSUE TITLE
Please provide hub for Ubuntu as Debian package NOT snap

### ISSUE THREAD

**Issue by ascheman** [github/hub#2434]
**The problem I'm trying to solve:**
Installing `hub` on an Ubuntu based Docker image is impossible, since it is based on Snap. Snap requires to run a local daemon `snapd` which is completely against the principles of Docker.

**How I imagine hub could expose this functionality:**
Just submit the Hub Debian package to Ubuntu and/or provide a Debian package repository for Hub to enable installation via the classical `apt-get` (or similar) tool.

**Comment by mislav** [comment]
Thanks for writing in and explaining the problem.

We (hub maintainers) do not publish nor maintain neither the Snap package nor the Windows/Fedora/Arch/FreeBSD/Debian packages. These are all provided by the community, with their issues tracked in their respective repos (if available).

With that in mind, I'm not sure that this issue belongs in this repository, because it's very likely that we are not going to publish nor maintain the Ubuntu package for apt-get. Perhaps someone from the community could step up and make it happen?

**Comment by mvastola** [comment]
@mislav would it be possible to at least add in the requisite files to run `dpkg-buildpackage` on this repo?

**Comment by mislav** [comment]
@mvastola 
1. Which would these files be?
2. I'm not sure if there was a point to that, since the debian package is likely going to be built by someone else in another repository?

**Comment by mvastola** [comment]
1. [Basically](https://www.debian.org/doc/manuals/maint-guide/dreq.en.html), a file that triggers the make command, a file that contains the metadata for the package, one for copyright info and one with a changelog (can be generated from the repo's releases with a script).

2. Well, firstly, I was looking and I couldn't find anyone at the moment who is maintaining a repo with the latest stable code in a debian package. But the idea is that if I clone the repo and build it myself, it's better to install it as a package so it's easily uninstallable, detects conflicts, etc.

I will admit that this does appear to /approach/ maintaining your own packages for various distros, but all that's really involved is writing a command that triggers the makefile. And chances are you already support this in practice, because if someone posts an issue saying "I can't compile on Ubuntu", that's still something that needs to be addressed.

**Comment by mislav** [comment]
> I will admit that this does appear to _approach_ maintaining your own packages for various distros, but all that's really involved is writing a command that triggers the makefile.

Thank you for explaining. This makes sense.

Rather than individual people building each their own package from this repo, wouldn't it be better that someone maintains a Debian package in a central location where people can easily install it from with their package managers?

If someone was already familiar with the process of building Debian packages, I'm guessing this should not be hard. I'm imagining there would be a repository that consists mostly of scripts that download a hub tarball for a specific release and either re-uses our pre-built binaries for Linux, or builds its own via `make bin/hub man-pages` (requires Go).

**Comment by mvastola** [comment]
> Rather than individual people building each their own package from this repo, wouldn't it be better that someone maintains a Debian package in a central location where people can easily install it from with their package managers?

Well, yeah, but the ideal entity to do this would be the owner of the project, so GitHub, but you sounded reticent to take on that responsibility. But yes, this should be done. To facilitate doing so, it would make sense to include the files necessary to build the package in this repo.

> If someone was already familiar with the process of building Debian packages, I'm guessing this should not be hard. I'm imagining there would be a repository that consists mostly of scripts that download a hub tarball for a specific release and either re-uses our pre-built binaries for Linux, or builds its own via `make bin/hub man-pages` (requires Go).

Well that's essentially what I'm suggesting. The debian build tools will do everything for you to produce the `.deb` file if you have a `./debian` subdirectory in the project.  Beyond that, the only script (actually a `Makefile`) you'd need would be `./debian/rules`, and that'd only be a couple of lines, which would trigger assorted debian build helpers and/or call the `make` command you noted.

**Comment by mislav** [comment]
> Well, yeah, but the ideal entity to do this would be the owner of the project, so GitHub, but you sounded reticent to take on that responsibility.

That's right. Since Debian is not the only package format out there, taking responsibility for packaging would mean that this repository should also handle Chocolatey, Scoop, RPM, Snap, FreeBSD and other package formats, and I feel that all of that would be hard to maintain over time.

The only package manager we officially support is Homebrew because it's 1) the one I use; and 2) [as simple as this](https://github.com/github/hub/blob/373d387eba9d7ef7780e54d2c828b0c0a8db28c4/.github/workflows/release.yml#L25-L28) to keep updated via release automation.

**Comment by mvastola** [comment]
> That's right. Since Debian is not the only package format out there, taking responsibility for packaging would mean that this repository should also handle Chocolatey, Scoop, RPM, Snap, FreeBSD and other package formats, and I feel that all of that would be hard to maintain over time.
> 
> The only package manager we officially support is Homebrew because it's 1) the one I use; and 2) [as simple as this](https://github.com/github/hub/blob/373d387eba9d7ef7780e54d2c828b0c0a8db28c4/.github/workflows/release.yml#L25-L28) to keep updated via release automation.

I get the argument there, but there are others that are just as simple. I agree that maintaining pre-packaged releases for different repos involves a bit of overhead (though GitHub does do this for `git-lfs` using packagecloud.io, with whom I have to assume there is some sort of agreement), but I think supporting build scripts would be pretty neutral in terms of ease of maintenance. 

The reason for this is, regardless of if this repo has distribution specific build scripts, it still needs to support building on all these different platforms. By including build scripts and instructing users to use them, a lot of potential issues are avoided because:
1. Regardless, the command(s) to build the program need to be documented in this repo; a build script just means putting these commands in a standardized location. If the build scripts aren't working, that will very often mean the equivalent commands that could be run manually aren't producing the expected result either. This needs to be addressed in any case.
2. Build scripts standardize the build process as well as do sanity checking in terms of ensuring dependencies are present (with the correct versions) and avoiding conflicts. The result is that everyone on Debian/Ubuntu simply needs to run `dpkg-buildpackage` inside the directory and nearly all of the guesswork is removed. (Conflicts and missing dependencies will be alerted, etc).
3. Using packages also facilitates clean uninstalls and upgrades of programs in a way that removes leftover files, thereby preventing some very-hard-to-debug issues from arising.
